include/igl/predicates/ear_clipping.cpp

```diff
@@ -0,0 +1,111 @@
+
+#include <igl/slice.h>
+#include "ear_clipping.h"
+#include "point_inside_polygon.h"
+#include "predicates.h"
+
+template <typename DerivedP, typename DerivedRT,
+          typename DerivedF, typename DerivedI>
+IGL_INLINE void igl::predicates::ear_clipping(
+  const Eigen::MatrixBase<DerivedP>& P,
+  const Eigen::MatrixBase<DerivedRT>& RT,
+  Eigen::PlainObjectBase<DerivedI>& I,
+  Eigen::PlainObjectBase<DerivedF>& eF,
+  Eigen::PlainObjectBase<DerivedP>& nP
+){
+  
+  // check whether vertex i is an ear
+  auto is_ear = [](
+    const Eigen::MatrixBase<DerivedP>& P,
+    const Eigen::MatrixBase<DerivedRT>& RT,
+    const Eigen::VectorXi& L,
+    const Eigen::VectorXi& R,
+    const int i
+  ){
+    int a = L(i), b = R(i);
```

Comment:
everywhere you are using D and I, you should use the correct type coming from Derived